cloudinit/util.py

```diff
@@ -1887,7 +1887,12 @@ def is_link(path):
 def sym_link(source, link, force=False):
     LOG.debug("Creating symbolic link from %r => %r", link, source)
     if force and os.path.lexists(link):
-        del_file(link)
+        # Provide atomic update of symlink to avoid races with status --wait
+        # LP: #1962150
+        tmp_link = os.path.join(os.path.dirname(link), "tmp" + rand_str(8))
+        os.symlink(source, tmp_link)
```

Comment:
what guarantees do you have that `tmp_link` doesn't exist on the filesystem?

Reply:
We have none, but the `rand_str(8)` makes it pretty unlikely. I suppose we could switch to a uuid if we want to be absolutely sure.

Reply:
am just thinking about the Bad Guy :tm: who pre-creates this and then gets you to use it

Reply:
Sorry, pre-creates what?

Reply:
Scenario where bad_guy brute force creates all tmp<suffix>  links before we get there results in a traceback in cloud-init which will stop cloud-init in it's tracks intead of silently allowing malicios data injection (even though our use-cases for symlink don't really lead to anything malicious at the moment).
```bash
python3 -c 'import os, pathlib; pathlib.Path("/file1").touch(); os.symlink("/file1", "badguytmpfile"); os.symlink("/run/cloud-initfile", "badguytmpfile")'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
FileExistsError: [Errno 17] File exists: '/run/cloud-initfile' -> 'badguytmpfile'
```

Reply:
I suppose just the go the extra mile and avoid the traceback in the event that we happen to have a collision:
```diff
diff --git a/cloudinit/util.py b/cloudinit/util.py
index ae3311e3e..df841f210 100644
--- a/cloudinit/util.py
+++ b/cloudinit/util.py
@@ -1890,6 +1890,7 @@ def sym_link(source, link, force=False):
         # Provide atomic update of symlink to avoid races with status --wait
         # LP: #1962150
         tmp_link = os.path.join(os.path.dirname(link), "tmp" + rand_str(8))
+        del_file(tmp_link)
         os.symlink(source, tmp_link)
         os.replace(tmp_link, link)
         return
```

That would guarantee we never see a Traceback because we clean up our tmp_link before we even attempt to create it,.

Reply:
If you're fine with the traceback (I don't mind, up to you folks) then think actually omitting the `del_file` is cleaner since you can't guarantee that Bad Guy:tm: won't create the symlink between us deleting it and then creating it (and also has a higher chance of discovering the destination because of the interaction with the FS).

In other words, let's land this! Thanks for helping to get it over the line

Reply:
+1 awesome thanks @sparkiegeek I  agree and will add the del_file. and land this.